src/librustc_trans/trans/cabi_x86_64.rs

```diff
@@ -410,18 +410,46 @@ pub fn compute_abi_info(ccx: &CrateContext,
         }
     }
 
-    let mut arg_tys = Vec::new();
-    for t in atys {
-        let ty = x86_64_ty(ccx, *t, |cls| cls.is_pass_byval(), Attribute::ByVal);
-        arg_tys.push(ty);
-    }
+    let mut avail_regs = 6; // RDI, RSI, RDX, RCX, R8, R9
```

Comment:
You still need to track them separately... suppose you have a struct which contains a `long long` and a `double`.  If you run out of xmm registers, the struct needs to be explicitly byval instead of being split up: otherwise, the integer part of the struct ends up in a register.  (clang has similar logic; see https://github.com/llvm-mirror/clang/blob/335cc9e22d8b977d83a12297fc85b6f6be965689/lib/CodeGen/TargetInfo.cpp#L2804 .)